### DISCUSSION TITLE
How to get historical weather data (nearly) free of charge?

### DISCUSSION THREAD

**Discussion by man-pro** [mytourbook/mytourbook#1061] [Ideas]
I want to get the historical weather data for all my previous tours of the last years. I think this would be a possibility: OpenWeather offers a "One Call by Call" subscription plan (https://openweathermap.org/price, 2023/4/12)
> •	**1,000 API calls per day for free**
> •	0.0014 EUR per API call over the daily limit
> 
> Make one API call and recieve all essential weather data in one response:
> •	Minute forecast for 1 hour
> •	Hourly forecast for 48 hours
> •	Daily forecast for 8 days
> •	**Historical data for 40+ years back by timestamp**
> •	National weather alerts
> 

For me it sounds great, I think it’s (nearly) free for someone who only wants to have the weather data for the current track of today and for all previous tours of the last years. Maybe you should not call up all the weather data of all previous the tours on one day. If you only fetch the data of significant less than 1.000 tours per day, then you should not exceed the free 1000 API calls per day.

For the historical data must be specified an API key, but unfortunately this is not implemented in the interface for OpenWeather yet.

I think it would be a great feature.

What do you think?

**Comment by FJBDev** [comment]
> I want to get the historical weather data for all my previous tours of the last years.

You can already do that in MyTourbook since I added the weather retrieval in 2019 ([19.7](https://mytourbook.sourceforge.io/mytourbook/index.php/history/history-2019#v19.7.0)).

If you get your own API key for [WorldWeatherOnline ](https://www.worldweatheronline.com/weather-api/api/pricing2.aspx)(like you say it's "nearly" free too ), you will be able to retrieve all the historical weather you want/need.
You can see the documentation [here](https://mytourbook.sourceforge.io/mytourbook/index.php/documentation/tools/documentation-tools-weather)

Then, later when I found them, I added the other vendors OpenWeatherMap and WeatherAPI for the users that didn't want to pay to retrieve the weather since, historically, MyTourbook is a program completely free of charge. And while those vendors are free, yes, they are limited to historical data for only several days (5 to 7)

The API 3.0 above you mention is not free and like WorldWeatherOnline, we will not ask the users to pay to use MyTourbook.

**Comment by man-pro** [comment]
I had looked at the WorldWeatherOnline interface. Unfortunately, the historical weather data from WorldWeatherOnline only goes back to the second half of 2009. 

Moreover, it is only free for the first 30 days, then it starts to get really expensive. Have a look to the _Pro Plan Calculater_ (https://www.worldweatheronline.com/weather-api/api/pricing-form.aspx)

> API Features Selected
> Local Weather API
> • Real Time & 3 day weather
> • Historical/Past Local Weather API 
> 
> API Requests per Day
> • 500 per day
> 
> Also Included
> • Time Zone API
> • Search API
> • Astronomy API
> • Support by Email
> • 99% up time
> 
> Not Included
> • SLA
> • Priority Support
> 
> Monthly: €55.55
> Yearly: €599.94

With OpenWeather, the first 1000 API calls per day are free, then you have to pay something. But 1000 calls are a lot. If you are careful, you will never pay anything. 

If you compare WorldWeatherOnline with OpenWeather, you get 500 API calls per day more with comparable features without having to pay anything. I think it would be a really good and useful extension. 


And one more idea: The textfield _Weather_ in the _Tour Editor_ is always filled with weather data. Unfortunately, this also deletes all your own entries. If you want to call the historical data, at first you have to copy your own content and after you have to paste it into the field back. This may be ok for a few tours, but ... It would be great if the contents of the field were preserved. 

Maybe you can "mark" your entries. Something like this. This way you could also enter the data of different providers.

> Sunny, warm day, a bit windy
> 
> --- mtb start / do not edit this section ---
> Retrieving weather data with "WorldWeatherOnline" 13/04/2023, 11:01:39 
> 25/09/2021, 11:11: ☀ Sunny, 22°C, max. 24°C, min. 20°C, feels like 23°C, 6km/h from S, 50% humidity, 1016.71mbar pressure 
> 11h 20°C feels like 20°C 6km/h from 193° humidity 59% pressure 1018.0mbar precipitation 0.0mm snowfall 0.0mm
> 12h 22°C feels like 22°C 6km/h from 193° humidity 53% pressure 1017.0mbar precipitation 0.0mm snowfall 0.0mm
> 13h 22°C feels like 25°C 6km/h from 192° humidity 50% pressure 1017.0mbar precipitation 0.0mm snowfall 0.0mm
> 14h 24°C feels like 25°C 6km/h from 191° humidity 47% pressure 1017.0mbar precipitation 0.0mm snowfall 0.0mm
> 15h 24°C feels like 25°C 6km/h from 191° humidity 44% pressure 1016.0mbar precipitation 0.0mm snowfall 0.0mm
> 16h 23°C feels like 24°C 5km/h from 182° humidity 47% pressure 1016.0mbar precipitation 0.0mm snowfall 0.0mm
> 17h 21°C feels like 21°C 4km/h from 174° humidity 50% pressure 1016.0mbar precipitation 0.0mm snowfall 0.0mm
> 
> Weather Station: "Friedberg", Distance from tour: 4.0km 
> --- mtb end ---
> 

It would be realy cool if you could use markdown internally for the textfields. Then you can mark the weather data internally and always exchange or expand it without changing the entries made by the user. As an option, you could also enable markdown formatting for users.

**Reply by FJBDev** [reply]
> With OpenWeather, the first 1000 API calls per day are free, then you have to pay something. But 1000 calls are a lot. If you are careful, you will never pay anything.

Maybe OWM is cheaper than WWO (today but pricing can change any day) but their API 3.0 still is not free. Today, the use of OWM is completely free for all the MT users. **If you are careful** ... but what if the MT users are not careful, then someone has to pay something, who will pay for it ? you ?


> And one more idea: The textfield _Weather_ in the _Tour Editor_ is always filled with weather data. Unfortunately, this also deletes all your own entries. If you want to call the historical data, at first you have to copy your own content and after you have to paste it into the field back. This may be ok for a few tours, but ... It would be great if the contents of the field were preserved.
> 
> Maybe you can "mark" your entries. Something like this. This way you could also enter the data of different providers.
> 

That is a possibility but something off topic for this discussion. Can you create a separate feature request ?


> It would be realy cool if you could use markdown internally for the textfields. Then you can mark the weather data internally and always exchange or expand it without changing the entries made by the user. As an option, you could also enable markdown formatting for users.

I am not sure that SWT supports markdown. Feel free to create a feature request too if you want because it's off topic to this current discussion

**Comment by man-pro** [comment]
> Today, the use of OWM is completely free for all the MT users.

What did I not see? I think the historical weather API ist only free for 30 days (https://www.worldweatheronline.com/weather-api/api/historical-weather-api.aspx). As already shown above, then it becomes really expensive.

> [...] pricing can change any day.

That's right. That applies to all providers. All providers offer the daily data free of charge, but if you want to retrieve **historical data**, it becomes expensive at OWM (~55€ a month).

The OW offer is currently extremely cheap (1,000 API calls per day for free / 0.0014 EUR per API call over the daily limit) and you only get the API key if you sign a contract. So you know what you are doing and price changes are communicated to you. I would like to access the historical data from OW with MT, even if I might have to pay a few cents. OW simply offers data going back further than just to mid-2009 and with much better conditions. 

1.000 calls a day, that's a lot. I think you can't achieve that even if you call up the historical data for all your tours on one day plus the tours of today. 

Your desire to offer only free services is great. Sometimes users are willing to take the risk of spending a few cents to get more historical data (my first track is from August 1998). It would be really great to get this opportunity.

**Reply by FJBDev** [reply]
> 1.000 calls a day, that's a lot. I think you can't achieve that even if you call up the historical data for all your tours on one day plus the tours of today.
> 
> Your desire to offer only free services is great. Sometimes users are willing to take the risk of spending a few cents to get more historical data (my first track is from August 1998). It would be really great to get this opportunity.

Maybe we won't achieve 1,000 calls per day for the MT users but there is something very important to keep in mind. While I don't expose MT's API key publicly, MT uses my custom build NodeJS hosted app to make those weather requests. This means that anyone could technically make a request through this app from anywhere and make that number explode. And no, I will not ask MT's users to pay for that "risk"

**Reply by FJBDev** [reply]
@man-pro 
See my code changes [here](https://github.com/mytourbook/mytourbook/pull/1326) for the upcoming version of MT where MT will now be using the v3 of the OpenWeatherMap API.

The good news is that the historical data goes back to 1979 and remains free for all the MT users.
However, note that the number of requests are limited to 1000 in order for this service to remain free for all.
